libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigationProvider.kt

```diff
@@ -0,0 +1,66 @@
+package com.mapbox.navigation.core
+
+import android.content.Context
+import com.mapbox.android.core.location.LocationEngine
+import com.mapbox.android.core.location.LocationEngineProvider
+import com.mapbox.android.core.location.LocationEngineRequest
+import com.mapbox.navigation.base.options.NavigationOptions
+
+/**
+ * Singleton responsible for ensuring there is only one MapboxNavigation instance.
+ */
+object MapboxNavigationProvider {
+    @Volatile
+    private var mapboxNavigation: MapboxNavigation? = null
+
+    /**
+     * Create MapboxNavigation with provided options.
+     * Should be called before [retrieve]
+     *
+     * @param context
+     * @param navigationOptions
+     * @param locationEngine
+     * @param locationEngineRequest
+     */
+    @JvmOverloads
+    @JvmStatic
+    fun create(
+        context: Context,
+        navigationOptions: NavigationOptions,
+        locationEngine: LocationEngine = LocationEngineProvider.getBestLocationEngine(context.applicationContext),
+        locationEngineRequest: LocationEngineRequest = LocationEngineRequest.Builder(1000L)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .build()
+    ): MapboxNavigation {
+        synchronized(MapboxNavigationProvider::class.java) {
+            mapboxNavigation?.onDestroy()
+            mapboxNavigation = MapboxNavigation(
+                context,
+                navigationOptions,
+                locationEngine,
+                locationEngineRequest
+            )
+
+            return mapboxNavigation!!
+        }
+    }
+
+    /**
+     * Retrieve MapboxNavigation instance.
+     * Should be called after [create]
+     */
+    @JvmStatic
+    fun retrieve(): MapboxNavigation {
+        var instance = mapboxNavigation
+        if (instance == null) {
+            synchronized(MapboxNavigationProvider::class.java) {
+                instance = mapboxNavigation
+                if (instance == null) {
```

Comment:
Why do we perform the instance check twice?

Reply:
because when we get into `synchronized` block it can be non-null anymore

Reply:
Why don't synchronize for the first check in this case?

Reply:
synchronization works slow, so to make it work faster we check `null` without syncronization. 
if it non-null we just return it. If it null -> go inside the slow block
it's a common singleton `double-null` check approach

Reply:
Today I learned! :+1: